### PR TITLE
JSON Geometry Parser for Nexus-style Geometry

### LINKED ISSUE #26213: StartLiveData Needs Instrument Deserialisation
The problems are highlighted [here](https://confluence.esss.lu.se/display/DAM/Instrument+Geometry)

If live streaming, the Geometry used (and hence the DR results) may be completely different from those if processing the equivalent nexus file, even though both the `FileWriter` and `Mantid:StartLiveData` work from the same Kafka stream.

It would be excellent to have this issue solved and demonstrated on the next V20 experiment in September, where we would also be demonstrating other live streaming aspect.

AFAIK, the Nexus Streamer does not provide the JSON schema we need, but it's on @matthew-d-jones TODO List. We would need to have this before proceeding.

We can either use, or be guided by the `NexusGeometryParser` in terms of how to intepret the JSON message into a working Instrument directly.

### PR DESCRIPTION
**Pull request by LamarMoore**
**Description of work.**

This PR concerns the implementation of a geometry parser for the Nexus kafka stream. The geometry will be in JSON format having the nexus file structure  (and classes) superimposed. 

**To test:**

- Through code review
- Ensure unit tests are sensible and pass.
- Ensure performance test is passes and performance is acceptable.

Fixes #26213 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/TestHelpers/src/JSONGeometryParserTestHelper.cpp [hunk 1]
```diff
@@ -0,0 +1,876 @@
...
+}
+
+void appendToChildren(Json::Value &parent, const Json::Value &child) {
+  Json::Value &children = parent["children"];
+  children[children.size()] = child;
+}
+
+void resizeValues(Json::Value &values, int32_t size) {
+  if (values.size() == 0) {
+    values.resize(size);
```
Comment (matthew-d-jones) [inline Framework/TestHelpers/src/JSONGeometryParserTestHelper.cpp]: Casting `size`, could use `size_t` as function arg or static_cast here.

File: Framework/TestHelpers/src/JSONGeometryParserTestHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,876 @@
...
+      } else
+        resizeValues(val, size);
+    }
+  }
+}
+
+template <class T>
+void fillValues(Json::Value &values, const std::vector<T> &fillArray,
+                int &start, int size) {
+  if (values.size() > 0) {
```
Comment (matthew-d-jones) [inline Framework/TestHelpers/src/JSONGeometryParserTestHelper.cpp]: `empty`

**Review by matthew-d-jones — CHANGES_REQUESTED** [review]
It looks like you are handling the "pixel_shape" case but not the "shape" case, where the geometry describes the entire instrument component. Let me know if there is not an example of that in the file I gave you.

I've also added some comments for minor improvements that could be made to the code.

**Review by OwenArnold — CHANGES_REQUESTED** [review]

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+  return getAllNXComponents(instrument, NXMONITOR);
+}
+
+std::vector<Json::Value> getAllChoppers(const Json::Value &instrument) {
+  return getAllNXComponents(instrument, NXDISK_CHOPPER);
+}
+
+template <class T>
+void addSingleValue(const Json::Value &val, std::vector<T> &fillArray) {
+  if (std::is_same<T, std::int64_t>::value)
```
[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: We aren't at C++ 17 yet. Will look into the `enable_if`

File: Framework/TestHelpers/src/JSONGeometryParserTestHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,876 @@
...
+
+void appendToChildren(Json::Value &parent, const Json::Value &child) {
+  Json::Value &children = parent["children"];
+  children[children.size()] = child;
+}
+
+void resizeValues(Json::Value &values, int32_t size) {
+  if (values.size() == 0) {
+    values.resize(size);
+    for (uint32_t i = 0; i < values.size(); ++i)
```
[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/TestHelpers/src/JSONGeometryParserTestHelper.cpp]: Can't. It breaks. I think the `operator[]` is resulting in the construction of an `Json::ArrayType`. When I take a reference to a `null` Json object and resize nothing happens. Might be a bug.

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: No, only c++14, though Martyn et al are working on getting c++17 support.

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+
+bool validateShapeInformation(const bool &isOffGeometry,
+                              const std::vector<Eigen::Vector3d> &vertices,
+                              const std::vector<uint32_t> &cylinders,
+                              const std::vector<uint32_t> &faces,
+                              const std::vector<uint32_t> &windingOrder) {
+  if ((isOffGeometry &&
+       (vertices.empty() || faces.empty() || windingOrder.empty())) ||
+      (!isOffGeometry && (vertices.empty() || cylinders.empty())))
+    return false;
```
Comment (OwenArnold) [inline Framework/NexusGeometry/src/JSONGeometryParser.cpp]: I'm not sure you should do it in this PR, but this seems like the information you might want to log. I changed `NexusGeometryParser` recently to allow logging via the `Kernel::Logger` maybe you could do the same thing here.

File: Framework/NexusGeometry/test/JSONGeometryParserTest.h [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+  }
+
+  void test_parse() {
+    TS_ASSERT_THROWS_NOTHING((JSONGeometryParser(instrument)));
+  }
+
+private:
+  std::string instrument;
+};
+
```
Comment (OwenArnold) [inline Framework/NexusGeometry/test/JSONGeometryParserTest.h]: Testing looks thorough 👍  for performance tests 👍  for so many destructive test cases

File: Framework/NexusGeometry/src/NexusGeometryParser.cpp [hunk 7]
```diff
@@ -30,12 +30,12 @@
...
 namespace Mantid {
 namespace NexusGeometry {
-
 using namespace H5;
 
 // Anonymous namespace
 namespace {
-
+const H5G_obj_t GROUP_TYPE = static_cast<H5G_obj_t>(0);
+const H5G_obj_t DATASET_TYPE = static_cast<H5G_obj_t>(1);
```
Comment (OwenArnold) [inline Framework/NexusGeometry/src/NexusGeometryParser.cpp]: These are unused?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Looking good. Few outstanding minorish changes IMHO.

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+const std::string VALUES = "values";
+
+bool validateNXAttribute(const Json::Value &attributes,
+                         const std::string &NXAttribute) {
+  auto NXAttr = attributes[0];
+  return (NXAttr[NAME] == NX_CLASS && NXAttr[VALUES] == NXAttribute);
+}
+
+Json::Value get(const Json::Value &entry, const std::string &NXAttribute) {
+  Json::Value item;
```
[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: `std::find_if` breaks things. Not sure why:
```cpp
  if(entry.empty())
      return Json::Value();
  
  auto item = std::find_if(
      entry.begin(), entry.end(), [&NXAttribute](const auto &child) {
        auto attributes = child[ATTRIBUTES];
        return (!attributes.isNull() &&
                 validateNXAttribute(attributes, NXAttribute));
      });
  return *item;
```

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+  boost::split(values, path, boost::is_any_of("/"), boost::token_compress_on);
+  std::reverse(values.begin(), values.end());
+  values.pop_back(); // remove empty first value
+  // Search dependency tree and verify all items present.
+  // removes value from vector once found.
+  recursiveDependencySearch(root[NEXUS_STRUCTURE], values);
+
+  // Left over values suggests the dependency could not be found
+  if (!values.empty())
+    throw std::invalid_argument("Could not find dependency in json provided.");
```
[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: Not sure I understand the question.

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+/// been found.
+void recursiveDependencySearch(const Json::Value &parent,
+                               std::vector<std::string> &values) {
+  auto name = parent[NAME].asString();
+  const auto &children = parent[CHILDREN];
+
+  if (!values.empty() && parent[NAME] == values.back())
+    values.pop_back();
+
+  if (children.isNull())
```
[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: Check is necessary but there is a better way to do it.

File: Framework/NexusGeometry/src/JSONGeometryParser.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: ```return *item``` won't be safe unless you can guarantee that the item will be found. `item` could be `entry.end()`

[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/NexusGeometry/src/JSONGeometryParser.cpp]: In that case won't `*item` return an empty `Json::Value`?

**Review by matthew-d-jones — APPROVED** [review]
Looks good to me, I can't spot any issues with the parsing now. It should be easier to spot any problems once we can ingest the JSON via Kafka and check things look correct in the instrument view.

**Review by OwenArnold — APPROVED** [review]

**Pull request merged**